File: ChatProject/Src/Server/LinkingContext.cpp
#include "ServerPCH.h"
#include "LinkingContext.h"

LinkingContext& LinkingContext::GetInstance()
{
	static LinkingContext instance;
	return instance;
}

NetworkId_t LinkingContext::GetNetworkId(shared_ptr<GameObject> gameObject)
{
	if (_gameObjectToNetworkIdMap.find(gameObject) == _gameObjectToNetworkIdMap.end())
	{
		return 0;
	}

	return _gameObjectToNetworkIdMap[gameObject];
}
shared_ptr<GameObject> LinkingContext::GetGameObject(NetworkId_t networkId)
{
	return _networkIdToGameObjectMap[networkId];
}
void LinkingContext::AddGameObject(shared_ptr<GameObject> gameObject)
{
	_networkIdToGameObjectMap[_nextNetworkId] = gameObject;
	_gameObjectToNetworkIdMap[gameObject] = _nextNetworkId;
	++_nextNetworkId;
}
void LinkingContext::RemoveGameObject(NetworkId_t networkId)
{
	auto gameObject = _networkIdToGameObjectMap[networkId];
	_networkIdToGameObjectMap.erase(networkId);
	_gameObjectToNetworkIdMap.erase(gameObject);
}

void LinkingContext::RemoveGameObject(shared_ptr<GameObject> gameObject)
{
	NetworkId_t networkId = _gameObjectToNetworkIdMap[gameObject];
	_networkIdToGameObjectMap.erase(networkId);
	_gameObjectToNetworkIdMap.erase(gameObject);
}

File: client/ChatToy/Assets/Scripts/Network/LinkingContext.cs
using System.Collections.Generic;
using UnityEngine;

public class LinkingContext
{
    public static LinkingContext Instance
    {
        get
        {
            if (_instance == null)
            {
                _instance = new LinkingContext();
            }
            return _instance;
        }
    }
    private static LinkingContext _instance;

    public void AddGameObject(uint networkId, GameObject gameObject)
    {
        _networkIdToGameObjectMap.Add(networkId, gameObject);
        _gameObjectToNetworkIdMap.Add(gameObject, networkId);
    }

    public uint GetNetworkId(GameObject gameObject)
    {


        return 0;
    }
    public GameObject GetGameObject(uint networkId)
    {
        if (_networkIdToGameObjectMap.TryGetValue(networkId, out GameObject gameObject) == false)
        {
            return null;
            
        }

        return gameObject;
    }

    private Dictionary<uint, GameObject> _networkIdToGameObjectMap;
    private Dictionary<GameObject, uint> _gameObjectToNetworkIdMap;

    private LinkingContext()
    {
        _networkIdToGameObjectMap = new Dictionary<uint, GameObject>();
        _gameObjectToNetworkIdMap = new Dictionary<GameObject, uint>();
    }
}
